CIAOS-Rust/src/storage.rs

```diff
@@ -0,0 +1,161 @@
+    }
+}
+
+
+// this function accepts the flatbuffer and returns the offset and size list after proess the files in it
+pub fn write_files_to_storage(body: &[u8])
+    -> Result<(Vec<u64>, Vec<u64>), Error> {
+    // Open the storage file "haystack.bin" in append mode
+    let mut haystack = OpenFile::new()?;
+
+    let mut offset_list: Vec<u64> = Vec::new();
+    let mut size_list: Vec<u64> = Vec::new();
+
+    //deserialises the binary data thats been accepted into FileDataList object using flatbuffer
+    
+    let file_data_list = match root::<FileDataList>(&body) {
+        Ok(data) => data,
+        Err(e) => return Err(ErrorBadRequest(format!("Failed to parse FlatBuffers data: {:?}", e))),
+    };
+    let files = match file_data_list.files() {
+        Some(files) => files,
+        None => return Err(ErrorBadRequest("No files found in FlatBuffers data")),
+    };
+    info!("Deserialized {} files", files.len());
+
+    for (index, file_data) in files.iter().enumerate() {
+        let data = match file_data.data() {
+            Some(data) => data,
+            None => {
+                error!("No data in file at index {}", index);
+                continue;
+            }
+        };
+
+        match haystack.write(data.bytes()) {
+            Ok((offset, size)) => {
+                offset_list.push(offset);
+                size_list.push(size);
+                info!("Written file {} at offset {} with size {}", index, offset, size);
+            }
+            Err(e) => {
+                error!("Failed to write file {} to haystack: {}", index, e);
+                return Err(ErrorInternalServerError(e));
+            }
+        }
+    }
+
+    Ok((offset_list, size_list))
+}
+
+
+pub fn get_files_from_storage(offset_list: Vec<u64>, size_list: Vec<u64>)-> Result<Vec<u8>, Error> {
+    info!("connecting to .bin and gettting files");
+    let mut haystack = OpenFile::new().map_err(ErrorInternalServerError)?;
+    info!("connected to .bin");
+    let mut builder = FlatBufferBuilder::new();
+    let mut file_data_vec = Vec::new();
+    info!("building the flatbuffer to share");
+    for (&offset, &size) in offset_list.iter().zip(size_list.iter()) {
+        let data = haystack.read(offset, size).map_err(ErrorInternalServerError)?;
+        let data_vector = builder.create_vector(&data);
+        let file_data = FileData::create(&mut builder, &FileDataArgs { data: Some(data_vector) });
+        file_data_vec.push(file_data);
+    }
+    info!("successfully built the buffer");
+    let files = builder.create_vector(&file_data_vec);
+    let file_data_list = FileDataList::create(&mut builder, &FileDataListArgs { files: Some(files) });
+    builder.finish(file_data_list, None);
+    info!("sending buffer");
+    Ok(builder.finished_data().to_vec())
+    }
+
+////////////////////////////////////////
+/// here starts delete functionality ///
+////////////////////////////////////////
+const DELETE_FILE: &str = "delete_log.json";
```

Comment:
JSON is not a good fit here. We'll have to use a technique called WAL(Write Ahead logging). Maybe we take it as a patch